File: typescript/src/bug_fixer.ts
export interface BugScenario {
    id: string;
    description: string;
    buggyCode: string;
    testCases: TestCase[];
    expectedFix: string;
}

export interface TestCase {
    id: string;
    input: any[];
    expectedOutput: any;
}

export interface TestResult {
    testName: string;
    passed: boolean;
    expected: any;
    actual: any;
    executionTimeMs: number;
}

export interface EvaluationResult {
    scenarioId: string;
    totalTests: number;
    passedTests: number;
    failedTests: number;
    results: TestResult[];
}

export class TSBugFixer {
    private scenarios: BugScenario[] = [];

    loadScenarios(): BugScenario[] {
        return this.scenarios;
    }

    analyze(code: string): { bugType: string; confidence: number } {
        const patterns: Record<string, string[]> = {
            "array-index": ["length", "[i]", "[arr.length]"],
            "async-error": ["async", "await", "Promise"],
            "type-coercion": ["==", "!=", "typeof"],
            "off-by-one": ["<=", ">=", "length - 1"],
            "closure-scope": ["let", "var", "const"],
        };

        let bestType = "unknown";
        let bestScore = 0;

        for (const [bugType, pats] of Object.entries(patterns)) {
            const score = pats.filter((p) => code.includes(p)).length;
            if (score > bestScore) {
                bestScore = score;
                bestType = bugType;
            }
        }

        return {
            bugType: bestType,
            confidence: Math.min(bestScore / 3, 1.0),
        };
    }

    suggestFix(code: string, bugType: string): string {
        return code;
    }

    testFix(
        originalCode: string,
        fixedCode: string,
        testCases: TestCase[]
    ): { original: EvaluationResult; fixed: EvaluationResult } {
        const original = this.validate(originalCode, testCases, "original");
        const fixed = this.validate(fixedCode, testCases, "fixed");
        return { original, fixed };
    }

    validate(
        code: string,
        testCases: TestCase[],
        scenarioId: string = "test"
    ): EvaluationResult {
        const results: TestResult[] = [];

        for (const tc of testCases) {
            const start = Date.now();
            try {
                const fn = new Function("return " + code)();
                const actual = fn(...tc.input);
                const elapsed = Date.now() - start;
                const passed = JSON.stringify(actual) === JSON.stringify(tc.expectedOutput);
                results.push({
                    testName: tc.id,
                    passed,
                    expected: tc.expectedOutput,
                    actual,
                    executionTimeMs: elapsed,
                });
            } catch (e: any) {
                results.push({
                    testName: tc.id,
                    passed: false,
                    expected: tc.expectedOutput,
                    actual: `Error: ${e.message}`,
                    executionTimeMs: Date.now() - start,
                });
            }
        }

        const passedCount = results.filter((r) => r.passed).length;
        return {
            scenarioId,
            totalTests: testCases.length,
            passedTests: passedCount,
            failedTests: testCases.length - passedCount,
            results,
        };
    }
}

export class TSTestRunner {
    private results: TestResult[] = [];

    runTests(
        code: string,
        testCases: { input: any[]; expected: any; name: string }[]
    ): TestResult[] {
        const results: TestResult[] = [];

        for (const tc of testCases) {
            const start = Date.now();
            try {
                const fn = new Function("return " + code)();
                const actual = fn(...tc.input);
                const elapsed = Date.now() - start;
                results.push({
                    testName: tc.name,
                    passed: JSON.stringify(actual) === JSON.stringify(tc.expected),
                    expected: tc.expected,
                    actual,
                    executionTimeMs: elapsed,
                });
            } catch (e: any) {
                results.push({
                    testName: tc.name,
                    passed: false,
                    expected: tc.expected,
                    actual: `Error: ${e.message}`,
                    executionTimeMs: Date.now() - start,
                });
            }
        }

        this.results = results;
        return results;
    }

    compareResults(results: TestResult[]): { passed: number; failed: number } {
        return {
            passed: results.filter((r) => r.passed).length,
            failed: results.filter((r) => !r.passed).length,
        };
    }

    generateReport(results: TestResult[]): string {
        const lines = ["=== TypeScript Test Report ===", ""];
        for (const r of results) {
            const status = r.passed ? "PASS" : "FAIL";
            lines.push(`[${status}] ${r.testName} - expected: ${r.expected}, actual: ${r.actual} (${r.executionTimeMs}ms)`);
        }
        const passed = results.filter((r) => r.passed).length;
        lines.push("");
        lines.push(`Passed: ${passed}/${results.length}`);
        return lines.join("\n");
    }
}

File: python/src/bug_fixer.py
"""Python Bug Fixer - analyzes and fixes buggy Python code."""

import json
import os
import sys
from typing import Any, Optional


class PythonBugFixer:
    """Analyzes and fixes buggy Python code scenarios."""

    def __init__(self):
        self.bug_scenarios: list[dict] = []

    def load_bug_scenarios(self, path: str) -> None:
        """Load bug scenarios from a JSON file."""
        if not os.path.isfile(path):
            return
        with open(path, "r") as f:
            self.bug_scenarios = json.load(f)

    def analyze_bug(self, code_snippet: str, error_trace: str = "") -> dict:
        """Analyze buggy Python code and identify the bug type.

        Returns:
            Dict with bug_type, description, and confidence.
        """
        indicators = {
            "off-by-one": ["range(len(", "range(1, len(", "while i < len("],
            "type_error": ["str(", "+ ''", "+ \"", "int("],
            "infinite_loop": ["while True", "while 1:", "while not"],
            "incorrect_operator": ["+ 1", "- 1", "!= ", "== False"],
            "null_reference": ["None", "not None", "if x:"],
            "missing_base_case": ["return", "def ", "if n =="],
        }

        scores = {}
        for bug_type, patterns in indicators.items():
            score = sum(1 for p in patterns if p in code_snippet)
            if error_trace:
                score += sum(1 for p in patterns if p in error_trace)
            scores[bug_type] = score

        best_type = max(scores, key=scores.get) if any(scores.values()) else "unknown"
        confidence = min(scores.get(best_type, 0) / 3.0, 1.0)

        return {
            "bug_type": best_type,
            "confidence": round(confidence, 2),
            "description": f"Detected {best_type.replace('_', ' ')} bug pattern",
        }

    def suggest_fix(self, code_snippet: str, bug_type: str) -> str:
        """Suggest a fix for the identified bug type."""
        fixes = {
            "off-by-one": code_snippet.replace(
                "range(1, len(", "range(1, len("
            ).replace("while i < len(", "while i <= len("),
            "type_error": code_snippet,
            "infinite_loop": code_snippet,
            "incorrect_operator": code_snippet,
            "null_reference": code_snippet,
            "missing_base_case": code_snippet,
        }
        return fixes.get(bug_type, code_snippet)

    def test_fix(
        self, original_code: str, fixed_code: str, test_cases: list[dict]
    ) -> dict:
        """Test the fix against test cases.

        Returns:
            Dict with original_results, fixed_results, and improvement.
        """
        original_results = self._run_tests(original_code, test_cases)
        fixed_results = self._run_tests(fixed_code, test_cases)

        orig_passed = sum(1 for r in original_results if r["passed"])
        fixed_passed = sum(1 for r in fixed_results if r["passed"])

        return {
            "original_passed": orig_passed,
            "original_total": len(test_cases),
            "fixed_passed": fixed_passed,
            "fixed_total": len(test_cases),
            "improvement": fixed_passed - orig_passed,
        }

    def validate_correctness(
        self, code_string: str, test_cases: list[dict]
    ) -> dict:
        """Validate code correctness against test cases.

        Returns:
            Dict with passed count, total, and details.
        """
        results = self._run_tests(code_string, test_cases)
        passed = sum(1 for r in results if r["passed"])
        return {
            "passed": passed,
            "total": len(test_cases),
            "all_pass": passed == len(test_cases),
            "details": results,
        }

    def _run_tests(self, code_string: str, test_cases: list[dict]) -> list[dict]:
        """Run code against test cases and capture results."""
        results = []
        for tc in test_cases:
            try:
                local_ns: dict = {}
                exec(code_string, {"__builtins__": __builtins__}, local_ns)
                func_name = None
                for key in local_ns:
                    if callable(local_ns[key]) and not key.startswith("_"):
                        func_name = key
                        break

                if func_name is None:
                    results.append({"test_id": tc.get("id", ""), "passed": False, "error": "No function found"})
                    continue

                result = local_ns[func_name](*tc.get("input", []))
                expected = tc.get("expected_output")
                results.append({
                    "test_id": tc.get("id", ""),
                    "passed": result == expected,
                    "result": result,
                    "expected": expected,
                })
            except Exception as e:
                results.append({
                    "test_id": tc.get("id", ""),
                    "passed": False,
                    "error": str(e),
                })
        return results
